File: limit/limit.nim
import math, strformat

proc prediksiLimit(f: proc(x: float): float, a: float, epsilon: float): tuple[kiri, kanan, rata: float] =
  let kiri = f(a - epsilon)
  let kanan = f(a + epsilon)
  let rata = (kiri + kanan) / 2.0
  return (kiri, kanan, rata)

proc f(x: float): float =
  if abs(x - 4.0) < 1e-9:
    raise newException(ValueError, "fungsi ini dibagikan dengan nol")
  return (x.pow(2) - 16.0) / (x - 4.0)
  
try:
  let a = 4.0
  let epsilon = 1e-6
  let (kiri, kanan, rata) = prediksiLimit(f, a, epsilon)

  echo &"Prediksi limit x -> {a:.4f} dari kiri {kiri:.4f}"
  echo &"prediksi limit x -> {a:.4f} dari kanan: {kanan:.4f}"
  echo &"estimasi hasil: {rata:.3f}"

except ValueError as error:
  echo "Error nih: ", error.msg

File: limit/limit.cpp
#include <cmath>
#include <exception>
#include <iomanip>
#include <iostream>
#include <stdexcept>

using MathFunc = double (*)(double);

void prediksi_limit(MathFunc f, double a, double epsilon, double &kiri,
                    double &kanan, double &rata) {
  kiri = f(a - epsilon);
  kanan = f(a + epsilon);
  rata = (kiri + kanan) / 2.0;
}

double f(double x) {
  if (std::abs(x - 4.0) < 1e-9) {
    throw std::runtime_error("dibagi dengan nol");
  }
  return (std::pow(x, 2) - 16.0) / (x - 4.0);
}

int main() {
  double a = 4.0;
  double epsilon = 1e-6;
  double kiri, kanan, rata;

  try {
    prediksi_limit(f, a, epsilon, kiri, kanan, rata);
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "prediksi limit saat x " << a << " dari kiri adalah: " << kiri
              << std::endl;
    std::cout << "prediksi limit saat x " << a
              << " dari kanan adalah: " << kanan << std::endl;
    std::cout << "hasilnya adalah: " << rata << std::endl;
  } catch (std::exception &error) {
    std::cerr << "error bang: " << error.what() << std::endl;
  }

  return 0;
}
